### PR TITLE
separate sqlwrite and sqlread server

### PR DESCRIPTION
**Pull request by H-M-H**
So, the idea is to separate the sql-servers, so you can specify different servers to read from and to write to. For example you could write to a distant masterserver and read from localhost (should replicate from master). Like this it is possible to have a few core/master-sql-servers and a whole replication-ring is no longer necessary.
Additionally this Pull Requests adds the possibility to specify more than one sql-server which makes it possible to fall back to another server if the first specified is unreachable.

In order to make all this stuff possible I restructured the sql-connection-stuff, we now have:

CSqlServer: This class provides access to one sql-server, connect and disconnect automatically lock the object to avoid several threads accessing it simultaneously.

CSqlConnector: This class is used to connect a sql-server, it will iterate over the list of given sql-servers and try to connect to one of them.

CSqlScore: This class now only contains things relevant to actually inserting records into db or querying some stuff. The sql-connection stuff has been moved to the above classes.

CSqlData: This struct contains the information relevant for the sql-query/statement.

CSqlExecData: This struct is used to call a function that does some sql-stuff within its own thread. It also supplies objects of CSqlData and CSqlServer to the called function.

### CODE REVIEW

**Comment by H-M-H** [comment]
@MyNewBie well it should be working, did not test it though.

**Comment by MyNewBie** [comment]
@H-M-H sql_server compiled done but get crash after running , cant connect to MySQL database.

**Comment by H-M-H** [comment]
This is very likely a problem with the sql library, the same thing happened on my mac. To fix it you could try:
1. compile sql libs on your own
2. check the compiler flags and additionally libs used for compiling sql libs and try to make them the same (on my mac for instance sql libs were compiled with -stdlib=libc++ but ddnet with -stdlib=libstdc++)

**Comment by H-M-H** [comment]
I guess I am done now, ofcourse this still needs extensive tests and possibly a review :).

**Comment by def-** [comment]
Any plan when this will be ready? Would be nice for the new Chile server, since it doesn't have enough RAM to run its own database. Would also like to get rid of a few other database servers, RUS and CHN especially.

**Comment by def-** [comment]
Any plan when this is ready? We're currently failing to store many ranks on Chile and Brazil because of DDoS attacks and other network problems.

**Comment by H-M-H** [comment]
It is nearly done, I will try to finish it this week.

**Comment by def-** [comment]
> Any plan when this is ready? We're currently failing to store many ranks on Chile and Brazil because of DDoS attacks and other network problems.

Still happens all the time, not sure why people even still play on these servers as their ranks are mostly not saved at all.

**Comment by def-** [comment]
What happens here when a server disappears while we've been using it? Will the write job be switched to another server and retried there? What if a map change happens meanwhile? Have you tested this? Maybe compile it to a special binary, and test it on a single server, for example Chile or Brazil, since they always have network problems anyway. If that works better than what we have right now, then you can add it to master.

**Comment by H-M-H** [comment]
What happens here when a server disappears while we've been using it?: The job indeed will be switched to the next specified server. If there is no server left inserts will be written to a file (if it is specified in the serverconfig).

Mapchanges while a sql-thread is running are probably still problematic.

**Comment by def-** [comment]
> Mapchanges while a sql-thread is running are probably still problematic.

Hm, that's bad. With the current approach the problem is that players change map right after finishing, so their finish never makes it from Chile to USA over the bad connection.

**Comment by H-M-H** [comment]
I think i can fix that, only /save would still be broken.

**Comment by H-M-H** [comment]
Okay, this fixes problems on mapchange/reload. New ranks will be saved even if the map is changed during the insert. Additionally some other unwanted behaviour is fixed (threads created before a mapchange notice the change and act accordingly now).

File: src/engine/server/server.h [hunk 1]
```diff
@@ -76,6 +81,12 @@ class CServer : public IServer
 	class IGameServer *m_pGameServer;
 	class IConsole *m_pConsole;
 	class IStorage *m_pStorage;
+
+#if defined (CONF_SQL)
+	CSqlServer* m_apSqlReadServers[MAX_SQLSERVERS];
+	CSqlServer* m_apSqlWriteServers[MAX_SQLSERVERS];;
```
Comment (def-) [inline src/engine/server/server.h]: double semicolon

File: autoexec_server.cfg [hunk 2]
```diff
@@ -195,3 +195,12 @@ sv_client_suggestion_old "Your DDNet client is old, update it on DDNet.tw!"
...
 sv_client_suggestion_bot "Your client has bots and can be remote controlled!\nPlease use another client like DDNet client from DDNet.tw"
+
+
+
+# CUSTOM CONFIG
+# -------------
+
+# if you do not want updates to overwrite your settings create a
+# file called myServerconfig.cfg and put your config there
+exec myServerconfig.cfg
```
Comment (def-) [inline autoexec_server.cfg]: Or we just don't update autoexec_server.cfg with the auto updater.

File: src/engine/server/sql_server.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+void CSqlServer::executeSql(const char *pCommand)
+{
+	m_pStatement->execute(pCommand);
+}
+
+void CSqlServer::executeSqlQuery(const char *pQuery)
+{
+	if (m_pResults)
+	delete m_pResults;
```
Comment (def-) [inline src/engine/server/sql_server.cpp]: indentation missing

File: src/engine/server/server.h (hunk 1 above)
Comment (H-M-H) [inline src/engine/server/server.h]: Good point, but there are more. Will fix this.

**Comment by def-** [comment]
What's with the `%%` in SaveScoreThread and SaveTeamScoreThread?

**Comment by def-** [comment]
What you're doing with SaveScoreThread and SaveTeamScoreThread looks dangerous. When you keep `%s` characters in code to use them later, it's possible that user-strings could contain them as well, no?

**Comment by H-M-H** [comment]
Right, i already thought about putting 'record' back instead of %s, but this assumes that every server is configured with 'record' as prefix. '%s' felt better but indeed not perfect. Any ideas to improve that ?

This commit introduced that btw: https://github.com/ddnet/ddnet/pull/378/commits/8dacd88755a20e4122c74d65248ce306cbec80cf .

File: autoexec_server.cfg (hunk 2 above)
Comment (H-M-H) [inline autoexec_server.cfg]: Yes but if we break compatibility with old config it is probably better to do it like this.

**Comment by def-** [comment]
This is being tested by HMH on Chile. He can merge it himself when he feels it's ready for general use.

**Comment by def-** [comment]
How are the Chile tests going?

**Comment by H-M-H** [comment]
Pretty good, already saved more than 100 ranks and found + fixed some bugs. Currently I am testing if fallbacks work fine with more than one sql-server.

**Pull request merged**